### PR TITLE
fix: Verify impls arising from function calls exist

### LINKED ISSUE #3469: Verify trait constraints arising from function calls
### Aim

`verify_trait_constraint` is only called for method calls currently, leading to trait constraints from other sources (variable references, and function calls namely) never being verified.

### Expected Behavior

`foo(a, b)` verifies any trait constraints required by foo are valid.

### Bug

If `foo(a, b)` requires a trait constraint with no matching impl, the compiler will panic during monomorphization.

### To Reproduce

1.
2.
3.
4.

### Installation Method

None

### Nargo Version

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves #3469

## Summary\*

Trait constraints are only validated currently on method calls. This means if a function call has a `where` constraint, it is never validated that an impl for it actually exists. This PR adds this check.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/type_check/expr.rs [hunk 1]
```diff
@@ -53,6 +53,18 @@ impl<'interner> TypeChecker<'interner> {
...
+
+                // Push any trait constraints required by this definition to the context
+                // to be checked later when the type of this variable is further constrained.
+                let definition = self.interner.definition(ident.id);
+                if let DefinitionKind::Function(function) = definition.kind {
+                    let function = self.interner.function_meta(&function);
+                    for mut constraint in function.trait_constraints.clone() {
+                        constraint.typ = constraint.typ.substitute(&bindings);
+                        self.trait_constraints.push((constraint, *expr_id));
+                    }
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/hir/type_check/expr.rs:65]:
```suggestion
                    for mut constraint in function.trait_constraints {
                        constraint.typ = constraint.typ.substitute(&bindings);
                        self.trait_constraints.push((constraint, *expr_id));
                    }
```

Comment (vezenovm) [reply compiler/noirc_frontend/src/hir/type_check/expr.rs:65]: Don't need to clone here

Comment (vezenovm) [reply compiler/noirc_frontend/src/hir/type_check/expr.rs:65]: Actually the clone is needed

**Review by kevaundray — APPROVED** [review]
Lets get this merged

**Pull request merged**